src/mds/MDSMap.cc

```diff
@@ -734,123 +680,141 @@ void MDSMap::sanitize(const std::function<bool(int64_t pool)>& pool_exists)
+    decode(last_failure, p);
+    decode(root, p);
+    decode(session_timeout, p);
+    decode(session_autoclose, p);
+    decode(max_file_size, p);
+    decode(max_mds, p);
+    decode(mds_info, p);
+    if (struct_v < 3) {
+      __u32 n;
+      decode(n, p);
+      while (n--) {
+	__u32 m;
+	decode(m, p);
+	data_pools.push_back(m);
+      }
+      __s32 s;
+      decode(s, p);
+      cas_pool = s;
+    } else {
+      decode(data_pools, p);
+      decode(cas_pool, p);
     }
-    __s32 s;
-    decode(s, p);
-    cas_pool = s;
-  } else {
-    decode(data_pools, p);
-    decode(cas_pool, p);
-  }
 
-  // kclient ignores everything from here
-  __u16 ev = 1;
-  if (struct_v >= 2)
-    decode(ev, p);
-  if (ev >= 3)
-    decode(compat, p);
-  else
-    compat = get_compat_set_base();
-  if (ev < 5) {
-    __u32 n;
-    decode(n, p);
-    metadata_pool = n;
-  } else {
-    decode(metadata_pool, p);
-  }
-  decode(created, p);
-  decode(modified, p);
-  decode(tableserver, p);
-  decode(in, p);
-  decode(inc, p);
-  decode(up, p);
-  decode(failed, p);
-  decode(stopped, p);
-  if (ev >= 4)
-    decode(last_failure_osd_epoch, p);
-  if (ev >= 6) {
-    if (ev < 10) {
-      // previously this was a bool about snaps, not a flag map
-      bool flag;
-      decode(flag, p);
-      ever_allowed_features = flag ? CEPH_MDSMAP_ALLOW_SNAPS : 0;
-      decode(flag, p);
-      explicitly_allowed_features = flag ? CEPH_MDSMAP_ALLOW_SNAPS : 0;
+    // kclient ignores everything from here
+    __u16 ev = 1;
+    if (struct_v >= 2)
+      decode(ev, p);
+    if (decode_ev)
+      ceph_assert(ev == 5);
+    if (ev >= 3)
+      decode(compat, p);
+    else
+      compat = get_compat_set_base();
+
+    decode(created, p);
+    decode(modified, p);
+    decode(in, p);
+    decode(inc, p);
+    decode(up, p);
+    decode(failed, p);
+    decode(stopped, p);
+    if (ev >= 7)
+      decode(inline_data_enabled, p);
+
+    if (ev >= 8) {
+      ceph_assert(struct_v >= 5);
+      decode(enabled, p);
+      decode(fs_name, p);
     } else {
-      decode(ever_allowed_features, p);
-      decode(explicitly_allowed_features, p);
+      if (epoch > 1) {
+	// If an MDS has ever been started, epoch will be greater than 1,
+	// assume filesystem is enabled.
+	enabled = true;
+      } else {
+	// Upgrading from a cluster that never used an MDS, switch off
+	// filesystem until it's explicitly enabled.
+	enabled = false;
+      }
     }
-  } else {
-    ever_allowed_features = 0;
-    explicitly_allowed_features = 0;
-  }
-  if (ev >= 7)
-    decode(inline_data_enabled, p);
 
-  if (ev >= 8) {
-    ceph_assert(struct_v >= 5);
-    decode(enabled, p);
-    decode(fs_name, p);
-  } else {
-    if (epoch > 1) {
-      // If an MDS has ever been started, epoch will be greater than 1,
-      // assume filesystem is enabled.
-      enabled = true;
+    if (ev >= 9) {
+      decode(damaged, p);
+    }
+
+    if (!decode_ev)
+      return;
+
+    // ev part
+    DECODE_START(1, p);
```

Comment:
you can't add DECODE_START here.

Reply:
add
```
if (devode_ev) {
DECODE_START
...
DECODE_END
}
``` 
at very end of this function

Reply:
That's what [this line](https://github.com/ceph/ceph/pull/33835/files#diff-c4d6a907104e07425dc624f7a66f572aR756) does, right?

Reply:
Please clarify

Reply:
no. don't nest the new DECODE_START/DECODE_END in old DECODE_START/DECODE_END

Reply:
no. don't nest the new DECODE_START/DECODE_END in old DECODE_START/DECODE_END

Reply:
It should be in a different scope `{}` then. I tried that with the below error.

```
/home/jcollin/workspace/ceph/src/mds/MDSMap.cc: In member function ‘void MDSMap::decode(ceph::buffer::v15_2_0::list::const_iterator&)’:
/home/jcollin/workspace/ceph/src/include/encoding.h:1358:8: error: redeclaration of ‘__u8 struct_v’
 1358 |   __u8 struct_v, struct_compat;      \
      |        ^~~~~~~~
/home/jcollin/workspace/ceph/src/mds/MDSMap.cc:767:5: note: in expansion of macro ‘DECODE_START’
  767 |     DECODE_START(1, p);
      |     ^~~~~~~~~~~~
/home/jcollin/workspace/ceph/src/include/encoding.h:1375:8: note: ‘__u8 struct_v’ previously declared here
 1375 |   __u8 struct_v;       \
      |        ^~~~~~~~
/home/jcollin/workspace/ceph/src/include/encoding.h:1457:3: note: in expansion of macro ‘__DECODE_START_LEGACY_COMPAT_LEN’
 1457 |   __DECODE_START_LEGACY_COMPAT_LEN(v, compatv, lenv, 1u, bl)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jcollin/workspace/ceph/src/mds/MDSMap.cc:690:3: note: in expansion of macro ‘DECODE_START_LEGACY_COMPAT_LEN_16’
  690 |   DECODE_START_LEGACY_COMPAT_LEN_16(5, 4, 4, p);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jcollin/workspace/ceph/src/include/encoding.h:1368:12: error: redeclaration of ‘unsigned int struct_end’
 1368 |   unsigned struct_end = bl.get_off() + struct_len;   \

```
After putting it in another scope, vstart still hangs at `fs volume create a`.

Reply:
push the code, so I can check it

Reply:
@ukernel pushed